californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java

```diff
@@ -284,24 +284,55 @@ public void receiveResponse(final Exchange exchange, final Response response) {
 		exchange.setFailedTransmissionCount(0);
 		exchange.setRetransmissionHandle(null);
 
-		if (response.getType() == Type.CON && !exchange.getRequest().isCanceled()) {
-			if (exchange.getSendNanoTimestamp() > response.getNanoTimestamp()) {
-				// received before ACK/RST was sent
-				int count = counter.incrementAndGet();
-				LOGGER.debug("{}: {} duplicate response {}, server sent ACK delayed, ignore response", count,
-						exchange, response);
-				return;
+		if (response.getType() == Type.CON) {
+			boolean ack = true;
+			if (response.isDuplicate()) {
+				if (response.getNanoTimestamp() < exchange.getSendNanoTimestamp()) {
+					// received response duplicate before ACK/RST
+					// or last request retransmission was sent
+					// => drop response
+					// Note: if the response is received the 1. time, no ACK nor RST was sent
+					// so far. Therefore the send timestamp is related to the request
+					// retransmission only. In that case always ACK/RST to try stopping future
+					// response retransmissions.
+					int count = counter.incrementAndGet();
+					LOGGER.info("{}: {} duplicate response {}, server sent ACK delayed, ignore response", count,
+							exchange, response);
+					return;
+				}
+				// resend last ack or rst, don't update, request state may have changed!
+				if (response.isRejected()) {
+					ack = false;
+					LOGGER.debug("{} reject duplicate CON response, request canceled.", exchange);
```

Comment:
Is the "request canceled" in the log intended ?

Reply:
The "canceled request" is the cause for the "rejected response".